File: Tailor/Geometry.py
from math import acos, sin, cos, degrees, sqrt, radians

def distance(a, b):
    return (a - b).length()

def angle(a, b, c):
    return degrees((b - a).angle(b - c))

def torsion(a, b, c, d):
    l = (b - a).cross(c - b)
    r = (d - c).cross(b - c)
    a = degrees(l.angle(r))

    if (l.cross(r) * (c - b)) < 0.0:
        return -a
    else:
        return a

def makeXYZFromAngle(a, b, distance, angle, plane_vector):
    angle_radians = radians(180 - angle)
    b_a = b - a
    b_a_normalised = b_a.normalize()
    x = distance * cos(angle_radians) * b_a_normalised
    y = distance * sin(angle_radians) * plane_vector
    v = x + y
    return b + v

def makeXYZ(previousXYZ, distance, previousButOneXYZ, angle, furthestXYZ, torsion):
    AB = furthestXYZ - previousButOneXYZ
    CB = previousXYZ - previousButOneXYZ
    normalAC = AB.cross(CB)

    if (normalAC.length() == 0.0):
        return makeXYZFromAngle(previousButOneXYZ, previousXYZ, distance, angle, Y)

    normalBC = CB.cross(normalAC)

    torsion_radians = radians(torsion)
    r = normalAC.normalize() * cos(torsion_radians) + normalBC.normalize() * sin(torsion_radians)
    newFrameOfReference = CB.normalize().cross(r)

    angle_radians = radians(angle)
    DC = -CB.normalize() * cos(angle_radians) + newFrameOfReference * sin(angle_radians)
    DC *= distance

    return previousXYZ + DC

def invertTorsion(torsion):
    if (torsion < 0):
        inverted_torsion = 180 + torsion
    else:
        inverted_torsion = torsion - 180
    return inverted_torsion

class Vector:

    def __init__(self, x, y, z):
        self.x = x
        self.y = y
        self.z = z

    def __repr__(self):
        return "(%.3f %.3f %.3f)" % (self.x, self.y, self.z)

    def __neg__(self):
        return Vector(-self.x, -self.y, -self.z)

    def __add__(self, other):
        if isinstance(other, Vector):
            return Vector(self.x + other.x, self.y + other.y, self.z + other.z)
        else:
            return Vector(self.x + other, self.y + other, self.z + other)

    def __radd__(self, other):
        return self.__add__(other)

    def __sub__(self, other):
        return Vector(self.x - other.x, self.y - other.y, self.z - other.z)

    def __rsub__(self, other):
        return Vector(other.x - self.x, other.y - self.y, other.z - self.z)

    def __mul__(self, other):
        if isinstance(other, Vector):
            return sum((self.x * other.x, self.y * other.y, self.z * other.z))
        else:
            return Vector(self.x * other, self.y * other, self.z * other)

    def __rmul__(self, other):
        return self.__mul__(other)

    def __div__(self, i):
        return Vector(self.x / i, self.y / i, self.z / i)

    def length(self):
        return sqrt(self * self)

    def normalize(self):
        self = self / self.length()
        return self

    def angle(self, other):
        try:
            c = (self * other) / (self.length() * other.length())
        except ZeroDivisionError, z:
            return 0
        return acos(max(-1., min(1., c)))

    def cross(self, other):
        return Vector(self.y * other.z - self.z * other.y,
                      self.z * other.x - self.x * other.z,
                      self.x * other.y - self.y * other.x)

    def copy(self):
        return Vector(self.x, self.y, self.z)

O = Vector(0, 0, 0)
X = Vector(1, 0, 0)
Y = Vector(0, 1, 0)
Z = Vector(0, 0, 1)

if __name__=="__main__":

        v1 = Vector(0, 0, 1)
        v2 = Vector(0, 1, 0)
        v3 = Vector(1, 0, 0)
        v4 = Vector(1, 1, 1)

        print(v1 + v2)
        print(v2 - v3)
        print(v1 * v3)
        print(v1 / 3)
        print(v2.length())
        print(v4.copy())
        print(v1.cross(v3))

        v4.normalize()
        print(v4)

        print(distance(v1, v2))
        print(angle(v1, v2, v3))
        print(torsion(v1, v2, v3, v4))

File: src/aigen/geometry/Geometry.java
package aigen.geometry;

public class Geometry {
	
	public static final Vector Y = new Vector(0, 1, 0);
	public static final Vector NEG_Y = new Vector(0, -1, 0);
  
    public static double distance(Vector a, Vector b) {
        return a.subtract(b).length();
    }
    
    public static double angle(Vector a, Vector b, Vector c) {
        return Math.toDegrees(b.subtract(a).angle(b.subtract(c)));
    }
    
    public static double torsion(Vector a, Vector b, Vector c, Vector d) {
        Vector l = b.subtract(a).cross(c.subtract(b));
        Vector r = d.subtract(c).cross(b.subtract(c));
        double ang = Math.toDegrees(l.angle(r));
        
        if (l.cross(r).dot(c.subtract(b)) < 0.0) {
            return -ang;
        } else {
            return ang;
        }
    }
    
    public static Vector makeXYZFromAngle(Vector a, Vector b, double distance, 
                                          double angle, Vector planeVector) {
        double angleRadians = Math.toRadians(180 - angle);
        Vector bA = b.subtract(a);
        Vector bANormalised = bA.normalize();
        Vector x = bANormalised.multiply(distance * Math.cos(angleRadians));
        Vector y = planeVector.multiply(distance * Math.sin(angleRadians));
        Vector v = x.add(y);
        return b.add(v);
    }
    
    public static Vector makeXYZ(Vector previousXYZ, double distance, 
                                 Vector previousButOneXYZ, double angle, 
                                 Vector furthestXYZ, double torsion) {
        Vector AB = furthestXYZ.subtract(previousButOneXYZ);
        Vector CB = previousXYZ.subtract(previousButOneXYZ);
        Vector normalAC = AB.cross(CB);
        
        if (normalAC.length() == 0.0) {
            return makeXYZFromAngle(previousButOneXYZ, previousXYZ, distance, angle, Vector.Y);
        }
        
        Vector normalBC = CB.cross(normalAC);
        
        double torsionRadians = Math.toRadians(torsion);
        Vector r = normalAC.normalize().multiply(Math.cos(torsionRadians))
                   .add(normalBC.normalize().multiply(Math.sin(torsionRadians)));
        Vector newFrameOfReference = CB.normalize().cross(r);
        
        double angleRadians = Math.toRadians(angle);
        Vector DC = CB.normalize().multiply(-Math.cos(angleRadians))
                    .add(newFrameOfReference.multiply(Math.sin(angleRadians)));
        DC = DC.multiply(distance);
        
        return previousXYZ.add(DC);
    }
    
    public static double invertTorsion(double torsion) {
        if (torsion < 0) {
            return 180 + torsion;
        } else {
            return torsion - 180;
        }
    }
    
    public static void main(String[] args) {
        Vector v1 = new Vector(0, 0, 1);
        Vector v2 = new Vector(0, 1, 0);
        Vector v3 = new Vector(1, 0, 0);
        Vector v4 = new Vector(1, 1, 1);
        
        System.out.println(v1.add(v2));
        System.out.println(v2.subtract(v3));
        System.out.println(v1.dot(v3));
        System.out.println(v1.divide(3));
        System.out.println(v2.length());
        System.out.println(v4.copy());
        System.out.println(v1.cross(v3));
        
        v4 = v4.normalize();
        System.out.println(v4);
        
        System.out.println(distance(v1, v2));
        System.out.println(angle(v1, v2, v3));
        System.out.println(torsion(v1, v2, v3, v4));
    }
}
